Remove redundant multipart decoding

alexpbrown opened the pull request:
**Checklist**
This pull request...

- [x] Fixes a bug
- [ ] Adds new functionality
- [ ] Updates documentation and/or examples
- [ ] Adds tests
- [ ] Is a **breaking** change -- existing code written for this library may need to be rewritten to work after these changes.

**Description**
Doing `.to_string().decode()` will work for text documents, but will break e.g. audio files with an error like:
```UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 466: invalid start byte```

From what research I could do, it looks there there is no encoding/decoding in the example code found at the [requests_toolbelt docs](https://pypi.org/project/requests-toolbelt/), and POST/PATCH `data` in `requests` is expecting a MultipartEncocer object.

I tested this change with file upload code in `examples/followup.py` using both audio and text files.

alexpbrown committed:
Remove redundant multipart decoding

Doing `.to_string().decode()` will work for text documents, but will break e.g. audio files with an error like:
```UnicodeDecodeError: 'utf-8' codec can't decode byte 0xff in position 466: invalid start byte```

There is no encoding/decoding in the example code found at the requests_toolbelt docs[0] 

Tested with the `examples/followup.py` example file upload with both audio and text files.

[0]https://pypi.org/project/requests-toolbelt/

breqdev commented:
Thanks for taking the time to submit a pull request!

Your change will work fine for uploading files in a followup message, but it causes issues when sending a file in the initial response. Try running `examples/message_object.py`, specifically the `/readme` command:

```
127.0.0.1 - - [12/Nov/2022 15:54:46] "POST /interactions HTTP/1.1" 500 -
Error on request:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/werkzeug/serving.py", line 319, in run_wsgi
    execute(self.server.app)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/werkzeug/serving.py", line 308, in execute
    for data in application_iter:
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/werkzeug/wsgi.py", line 462, in __next__
    return self._next()
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/werkzeug/wrappers/response.py", line 50, in _iter_encoded
    for item in iterable:
TypeError: 'MultipartEncoder' object is not iterable
```

You're correct in that passing the `MultipartEncoder` directly to `requests` works fine. However, sending the initial response requires passing this to Flask, which isn't able to handle a `MultipartEncoder`.

alexpbrown committed:
Update message.py

to_string() runs read() on the files and generates a mutipart body that should work with both Flask and Requests

alexpbrown commented:
Interesting, good catch! 

I spent a while trying to find a good way to write tests for this today, but didn't have much luck sadly :( The major problem being that `boundary` is autogenerated so it's not straightforward to `assert` a string match without re-parsing everything. The other problem of course is that I don't really know what I'm doing haha.

However, changing it to return `multipart.to_string()` works in both the followup and initial example cases as far as I can tell.

breqdev committed:
Make Message.encode() always return bytes

breqdev commented:
Yep, that seems to work! For consistency, I just pushed a change that makes the method always return bytes (by `.encode()`-ing the JSON data in the non-multipart case).

Thanks for taking the time to make this PR! If all goes well I'll push out a release with this update shortly.

alexpbrown commented:
Awesome, thanks for checking this out so quickly!
